File: utils/extract_strings.pl
#!/usr/bin/env perl
use strict;
use warnings;

use Getopt::Long;

my @requested_targets;
# The following are legal alternatives to the default string
my @targets = ("devo8", "devo10", "devo12");
my $PO_LANGUAGE_STRING = "->Translated Language Name<-";

sub main {
    my $update;
    my $lang;
    my $fs;
    my $target_list;
    my $objdir;
    my $elffile;
    my $count;
    my $po;
    my $format = "v1";

    $ENV{CROSS} ||= "";
    GetOptions("update" => \$update, "language=s" => \$lang, "fs=s" => \$fs,
        "targets=s" => \$target_list, "count" => \$count, "po" => \$po,
        "objdir=s" => \$objdir, "elffile=s" => \$elffile,
        "format=s" => \$format);
    @requested_targets = split(/,/, $target_list) if($target_list);
    check_targets($fs, @requested_targets);

    my $uniq = get_strings($elffile || $objdir);
    if($count) {
        printf "%d", scalar(keys(%$uniq));
        exit 0;
    }

    if(! $update) {
        if ($po) {
            print "msgid \"$PO_LANGUAGE_STRING\"\nmsgstr \"\"\n\n";
        }
        foreach (@{ $uniq->{__ORDER__} }) {
            if ($po) {
                # Match same syntax used by getopt
                print $uniq->{$_}; # Comment
                my @msg = split(/(?<=\\n)/, $_);
                if (scalar(@msg) > 1) {
                    print "msgid \"\"\n\"" . join("\"\n\"", @msg) . "\"\n";
                } else {
                    print "msgid \"$msg[0]\"\n";
                }
                print "msgstr \"\"\n";
                print "\n";
            } else {
                print ":$_\n";
           }
        }
        exit 0;
    }

    my @files = ($po ?
                 ($lang ?
                     "fs/language/locale/deviation.$lang.po" :
                     glob("fs/language/locale/*.po")) :
                 glob($lang ?
                     "fs/language/lang*.$lang" :
                     "fs/language/lang*"));
    for my $file (@files) {
        my($ext, $language, $translation) = $po
            ? parse_po_file($file, $uniq)
            : parse_v1lang_file($file, $uniq);
        write_lang_file("$fs/lang.$ext", $language, $translation, $format);
    }
}

sub check_targets {
    my($fs, @requested_targets) = @_;
    if($fs || @requested_targets) {
        if (! @requested_targets) {
            print "ERROR: Must specify both -targets with -fs\n";
            exit 1;
        }
        my $ok = 1;
        foreach my $t (@requested_targets) {
            if(! grep {$_ eq $t} @targets) {
                print "ERROR: Target '$t' is not in list (@targets)\n";
                $ok = 0;
            }
        }
        exit 1 if(! $ok);
    }
}

sub get_strings {
    my($path) = @_;
    my $strings = extract_all_strings();
    if ($path) {
        $strings = extract_target_strings($path, $strings);
    }
    #append template names
    foreach (`head -n 1 fs/common/template/*.ini`) {
        chomp;
        if(/template=(.*?)\s*$/) {
            push @{ $strings->{__ORDER__} }, $1 if (! $strings->{$1});
            $strings->{$1} = "#: Model template\n";
        }
    }
    return $strings;
}

sub extract_all_strings {
    my @lines = `/usr/bin/find . -name "*.[hc]" | grep -v libopencm3 | sort | xargs xgettext -o - --omit-header -k --keyword=_tr --keyword=_tr_noop --no-wrap`;
    my $idx = 0;
    my %strings = (__ORDER__ => [] );  # Crude Tie::IxHash implementation
    #Read all strings and put into a hash that maps the containing file to the string
    while(@lines) {
        my($msgid, $msgstr, $comment) = parse_gettext(\@lines);
        next unless($msgid);
        push @{ $strings{__ORDER__} }, $msgid if(! defined $strings{$msgid});
        $strings{$msgid} = $comment;
    }
    return \%strings;
}

sub extract_target_strings {
    my($path, $valid_strings) = @_;
    my %strings;
    my @files = (-d $path ? glob("$path/*.o") : $path);

    foreach my $file (@files) {
        #Parse all strings from the object files and add to the allstr hash
        my @od = `$ENV{CROSS}objdump -s $file`;
        my $str = "";
        my $state = 0;
        foreach(@od) {
            my $orig = $_;
            if(/section (\.rel)?\.ro?data/ || /section __cstring/) {
                $str = "";
                $state = 1;
                next;
            } elsif(/^Contents/ && ! /\.ro?data/) {
                $str = "";
                $state = 0;
            } elsif($state) {
                #Strip off everything but hex data
                s/^\s+\S+\s//;
                s/\s\s.*//;
                s/ //g;
                while(/(\S\S)/g) {
                    #Iterate over each ascii character
                    if($1 eq "00") {
                        #NULL termination
                        if($str) {
                            $str =~ s/\n/\\n/g;  #Convert <CR> to \n
                            $strings{$str} = $valid_strings->{$str} if($valid_strings->{$str});
                            $str = "";
                        }
                    } else {
                        $str .= chr(hex($1));
                    }
                }
            }
        }
    }
    $strings{__ORDER__} = [ grep { defined $strings{$_} } @{ $valid_strings->{__ORDER__} } ];
    return \%strings;
}

sub fnv_16 {
    my ($input, $init_value) = @_;

    $init_value = 0x811c9dc5 unless (defined $init_value);
    my $fnv_32_prime = 0x01000193;

    my $hval = $init_value;

    foreach my $x (unpack ('C*', $input)) {
        $hval = ($hval * $fnv_32_prime) & 0xffffffff;
        $hval = ($hval ^ $x) & 0xffffffff;
    }

    return ($hval >> 16) ^ ($hval & 0xffff);
}


sub write_lang_file {
    my($outf, $language, $translation, $format) = @_;
    my %strings;
    my %final;
    my %hashvalues;
    foreach ("DEFAULT", @requested_targets) {
        #Hierarchically try to find best string
        next unless($translation->{$_});
        %strings = (%strings, %{$translation->{$_}});
    }
    foreach (sort keys %strings) {
        next if ($_ eq $strings{$_});
        my $hash = fnv_16($_);
        my $value = $strings{$_};
        if (defined($hashvalues{$hash})) {
            printf("Error: Conflict hash detected:\n%s\n%s\n",
                   $hashvalues{$hash}, $value);
            exit 1;
        }
        $hashvalues{$hash} = $value;
        $final{$_} = $value;
    }

    open(my $fh, ">", $outf) or die "ERROR: Can't write $outf\n";
    print $fh $language;
    if ($format eq "v2") {
        foreach(sort{$a <=> $b} keys %hashvalues)
        {
            print $fh pack("CC", $_ & 0xFF, $_ >> 8);
            print $fh "$hashvalues{$_}\n";
        }
    } else {
       foreach (sort keys %final) {
           print $fh ":$_\n$final{$_}\n";
       }
    }
    close $fh;
}

sub parse_gettext {
    my($lines) = @_;
    my $msgid;
    my $comment = "";
    while(1) {
        my $line = shift(@$lines);
        if ($line =~ /^#[:,]/) {
            $comment .= $line;
            next;
        }
        if ($line =~ /^\s*(msgid|msgstr)\s+"(.*)"\s*$/) {
            my($type, $str) = ($1, $2);
            while(@$lines && $lines->[0] =~ /^\s*"(.*)"\s*$/) {
                $str .= $1;
                shift(@$lines);
            }
            if ($type eq "msgid") {
                if ($msgid) {
                    print "Error: msgid '$msgid' missing msgstr.  Ignoring\n";
                }
                $msgid = $str;
            }
            elsif ($type eq "msgstr") {
                if (! defined $msgid) {
                    print "Error: no msgid for msgstr '$str'.  Ignoring\n";
                    next;
                }
                return($msgid, $str, $comment || "#");
            }
        }
    }
}

sub parse_po_file {
    my($file, $uniq) = @_;
    my %strings;
    my $language = "Unknown";
    my $ext;
    my $re_target = join("|", @targets);

    open my $fh, "<", $file;
    my @lines = <$fh>;
    close($fh);
    while(@lines) {
        my($msgid, $msgstr) = parse_gettext(\@lines);
        $msgstr =~ s/\\([^n])/$1/g; # Fix escaped characters
        next if(! $msgstr);
        if ($msgid eq "" && $msgstr =~ /(?:\b|\\n)Language:\s+(\S+?)(?:\b|\\n)/) {
            $ext = lc($1);
            if (length($ext) > 3) {
                $ext =~ s/^.*_//;
                $ext = substr($ext, -3);
            }
        }
        next if(! $msgid);
        if ($msgid eq $PO_LANGUAGE_STRING) {
            # Add UTF-8 BOM
            $language = chr(0xef) . chr(0xbb) . chr(0xbf) . $msgstr . "\n";
            next;
        }
        next if(! $uniq->{$msgid});
        my @values = split(/\\n(?=(?:$re_target)?:)/, $msgstr);
        foreach (@values) {
            my($target, $target_str) = (/^(?:($re_target):)?(.*)/);
            $strings{$target}{$msgid} = $target_str if($target);
            $strings{DEFAULT}{$msgid} ||= $target_str;
        }
    }
    return ($ext, $language, \%strings);
}

sub parse_v1lang_file {
    my($file, $uniq) = @_;
    my %strings;
    open my $fh, "<", $file;
    my $language = <$fh>;
    my(@lines) = <$fh>;
    my ($ext) = ($file =~ /[^\/.]+\.([^\/.\s]+)(?:\.po)?$/);
    for (my $line = 0; $line < $#lines; $line++) {
        $_ = $lines[$line];
        chomp;
        if(/^:(.*)/) {               #/^([<>]?):(.*)/)
            my($eng) = ($1);
            next if ($line < $#lines-2 && $lines[$line+1] =~ /^:./);
            my $next = $lines[++$line];
            chomp $next;
            $strings{DEFAULT}{$eng} = $next if($uniq->{$eng});
        } elsif(/^\|(\S+?):(.*)/) {
            my($t, $eng) = ($1, $2);
            my $next = $lines[++$line];
            chomp $next;
            next if(! $uniq->{$eng});
            $strings{$t}{$eng} = $next;
            $strings{DEFAULT}{$eng} ||= $next;
        }
    }
    return($ext, $language, \%strings);
}

main();

# | xargs xgettext -o - --omit-header -k --keyword=_tr --keyword=_tr_noop --no-wrap | grep -v 'msgstr ""' | grep -v "^#" | grep -v "^$"

#: pages/dialogs.c:100
#msgid ""
#"Binding is in progress...\n"
#"Make sure model is on!\n"
#"\n"
#"Pressing OK will NOT cancel binding procedure\n"
#"but will allow full control of Tx."
#msgstr ""

File: utils/extract_strings.py
#!/usr/bin/env python3
"""Extract translatable strings out of object files and
   compare/convert to po translation files or Deviation language files
"""

import argparse
import os
import sys
import re
import subprocess
import logging
import glob

PO_LANGUAGE_STRING = "->Translated Language Name<-"
TARGETS = ["devo8", "devo10", "devo12"]
CROSS = os.environ.get("CROSS", "")
os.environ['LANG'] = 'en_US.UTF-8'  # We need English to be able to regex parse cmd output

def main():
    """Main routine"""
    parser = argparse.ArgumentParser()
    parser.add_argument("-update", action="store_true",
                        help="Update existing po files with modified string lists")
    parser.add_argument("-language",
                        help="Language or pattern of language files to update")
    parser.add_argument("-fs",
                        help="Directory of output language files")
    parser.add_argument("-targets", nargs='+', choices=TARGETS,
                        help="Specify targets to generate")
    parser.add_argument("-count", action="store_true",
                        help="Generate count of unique srings")
    parser.add_argument("-po", action="store_true",
                        help="Read po files rather than devo-formatted files")
    parser.add_argument("-objdir",
                        help="Directory containing .o files to extract strings from")
    args = parser.parse_args()

    if args.fs and not args.targets:
        logging.error("Must specify both -targets with -fs")
        return False

    uniq = get_strings(args.objdir)

    if args.count:
        print("{}".format(len(uniq.keys())))
        return True

    if not args.update:
        if args.po:
            print_po_strings(uniq)
        else:
            print_strings(uniq)
        return True
    files = (["fs/language/locale/deviation.{}.po".format(args.language)] if args.language
             else glob.glob("fs/language/locale/*.po"))
    for filename in files:
        (ext, language, translation) = parse_po_file(filename, uniq)
        if not write_lang_file("{}/lang.{}".format(args.fs, ext),
                               args.targets, language, translation):
            return False
    return True


def print_po_strings(uniq):
    """Generate po file containing all translatable strings"""
    print('msgid "{}"\nmsgstr ""\n'.format(PO_LANGUAGE_STRING))
    for string in uniq['__ORDER__']:
        # Match same syntax used by getopt
        print('\n'.join(uniq[string]))  # Comment
        # we need to split on '\\n' but preserver it as part of the previous line
        # python doesn't offer 'split' based on zero-length look-behind
        msg = [x for x in string.replace('\\n', '\\n\n').split('\n') if x]
        if len(msg) > 1:
            print('msgid ""\n"{}"'.format('"\n"'.join(msg)))
        else:
            print('msgid "{}"'.format(msg[0]))
        print('msgstr ""')
        print('')


def print_strings(uniq):
    """Generate deviation lang file containing all translatable strings"""
    for string in sorted(uniq['__ORDER__']):
        print(":{}".format(string))


def get_strings(path):
    """Get the full list of translatable strings, potentially filtered by target"""
    strings = extract_all_strings()
    if path:
        strings = extract_target_strings(path, strings)
    # append template names
    _re = re.compile(r'template=(.*?)\s*$')
    for line in system("head -n 1 fs/common/template/*.ini").split('\n'):
        _m = _re.search(line)
        if _m:
            if _m.group(1) not in strings:
                strings['__ORDER__'].append(_m.group(1))
            strings[_m.group(1)] = ["#: Model template"]
    return strings


def extract_all_strings():
    """Extaract full list of translatable strings from source-code"""
    strings = {"__ORDER__": []}
    cmd = ("find . -name '*.[hc]' | grep -v libopencm3 | sort "
           "| xargs xgettext -o - --omit-header -k --keyword=_tr "
           "--keyword=_tr_noop --no-wrap")
    _p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    while True:
        msgid, _msgstr, comment = parse_gettext(_p.stdout)
        if msgid is None:
            break
        if not msgid:
            continue
        if msgid not in strings:
            strings['__ORDER__'].append(msgid)
        strings[msgid] = comment
    return strings


def extract_target_strings(path, valid_strings):
    """Extract all strings from target object files"""
    strings = {}

    cmd = CROSS + "objdump -s " + os.path.join(path, "*.o")
    _p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    _str = ""
    state = 0
    for line in _p.stdout:
        line = line.decode('utf-8')
        if not line.startswith(' '):
            state = 0
            _str = ""
            if line.startswith("Contents") and any(x in line for x in [
                    "section .rel.rdata", "section .rel.rodata",
                    "section .rdata", "section .rodata", "section __cstring"]):
                state = 1
            continue
        if not state:
            continue
        line = re.sub(r'^\s+\S+\s', r'', line, 1)
        line = re.sub(r'\s\s.*', r'', line, 1)
        line = re.sub(r'\s', '', line)
        for _ch in re.findall(r'\S\S', line):
            if _ch == "00":
                # NULL termination
                if _str:
                    _str = _str.replace('\n', '\\n')
                    if _str in valid_strings:
                        strings[_str] = valid_strings[_str]
                    _str = ""
                continue
            _str += chr(int(_ch, 16))
    strings['__ORDER__'] = [_x for _x in valid_strings['__ORDER__'] if _x in strings]
    return strings


def fnv_16(data, init_value=0x811c9dc5):
    """Generate a FNV32 hash, and fold it into a 16bit value"""
    data = data.encode('utf-8')
    fnv_32_prime = 0x01000193
    hval = init_value

    for _ch in data:
        hval = (hval * fnv_32_prime) & 0xffffffff
        hval = (hval ^ _ch) & 0xffffffff
    return (hval >> 16) ^ (hval & 0xffff)


def write_lang_file(outf, targets, language, translation):
    """Write Deviation lang file for selected language"""
    strings = {}
    hashvalues = {}

    for target in ["DEFAULT"] + targets:
        # Hierarchically try to find best string
        if target in translation:
            strings.update(translation[target])
    for string in sorted(strings.keys()):
        value = strings[string]
        if string == value:
            continue
        hval = fnv_16(string)
        if hval in hashvalues:
            logging.error("Conflict hash detected:\n%s\n%s",
                          hashvalues[hval], value)
            return False
        hashvalues[hval] = value
    try:
        with open(outf, "wb") as _fh:
            _fh.write(language.encode('utf-8'))
            for key in sorted(strings.keys()):
                value = strings[key]
                if key == value:
                    continue
                _fh.write(":{}\n{}\n".format(key, value).encode('utf-8'))
    except OSError:
        logging.error("Can't write %s", outf)
        return False
    return True


def parse_gettext(_fh):
    """Parse next gettext element for string list"""

    msgid = None
    comment = []
    in_multiline = False
    re_msg = re.compile(r'^\s*(msgid|msgstr)\s+"(.*)"\s*$')
    re_quoted = re.compile(r'^\s*"(.*)"\s*$')
    _str = ""
    _type = ""
    # import pdb; pdb.set_trace()
    while True:
        line = _fh.readline().decode('utf-8')
        if in_multiline:
            _m = re_quoted.search(line)
            if _m:
                _str += _m.group(1)
                continue
            in_multiline = False
            if _type == "msgid":
                if msgid:
                    logging.error("msgid '%s' missing msgstr.  Ignoring", msgid)
                msgid = _str
            elif _type == "msgstr":
                if msgid is not None:
                    return(msgid, _str, comment or "#")
                logging.error("no msgid for msgstr '%s'.  Ignoring", _str)
        if not line:
            break
        if line.startswith('#:') or line.startswith('#,'):
            comment.append(line.rstrip())
            continue
        _m = re_msg.search(line)
        if _m:
            _type = _m.group(1)
            _str = _m.group(2)
            in_multiline = True
    return (None, None, None)


def parse_po_file(filename, uniq):    # pylint: disable=too-many-branches
    """Parse .po file into dictionary"""
    strings = {'DEFAULT': {}}
    language = "Unknown"
    ext = None
    re_target = "|".join(TARGETS)

    _fh = open(filename, "rb")
    while True:
        msgid, msgstr, _comment = parse_gettext(_fh)
        if msgid is None:
            break
        if msgstr:
            msgstr = re.sub(r'\\([^n])', r'\1', msgstr)  # Fix escaped characters
        if not msgstr:
            continue
        if msgid == "":
            _m = re.search(r'(?:\b|\\n)Language:\s+(\S+?)(?:\b|\\n)', msgstr)
            if _m:
                ext = _m.group(1).lower()
                if len(ext) > 3:
                    ext = re.sub(r'^.*_', r'', ext)
                    ext = ext[-3:]
            continue
        if msgid == PO_LANGUAGE_STRING:
            # Add UTF-8 BOM
            language = '\ufeff' + msgstr + "\n"
            continue
        if msgid not in uniq:
            continue

        values = re.split(r'\\n(?=(?:{})?:)'.format(re_target), msgstr)
        for value in values:
            _m = re.search(r'(?:({}):)?(.*)'.format(re_target), value)
            target = _m.group(1)
            target_str = _m.group(2)
            if target:
                if target not in strings:
                    strings[target] = {}
                strings[target][msgid] = target_str
            if msgid not in strings['DEFAULT']:
                strings['DEFAULT'][msgid] = target_str
    return (ext, language, strings)


def system(cmd):
    """Wrapper to easily call system functions"""
    logging.debug("Running: %s", " ".join(cmd))
    if isinstance(cmd, list):
        return subprocess.check_output(cmd).decode('utf-8').rstrip()
    return subprocess.check_output(cmd, shell=True).decode('utf-8').rstrip()

sys.exit(0 if main() else 1)
